File: Benchmarks.FSharp/FailureComparison.fs
namespace Benchmarks

open System.Text.Patterns
open FParsec
open BenchmarkDotNet.Attributes
open BenchmarkDotNet.Diagnosers

[<ClrJob>]
[<MemoryDiagnoser>]
type FailureComparison() =
    let pattern = Pattern.Number * 3 >> '-' >> Pattern.Number * 3 >> '-' >> Pattern.Number * 4;
    let parsec = attempt (digit .>>. digit .>>. digit .>>. pchar '-' .>>. digit .>>. digit .>>. digit .>>. pchar '-' .>>. digit .>>. digit .>>. digit .>>. digit)

    [<Benchmark>]
    member this.PatternFail() = pattern.Consume("555-555-Narwhal");

    [<Benchmark>]
    member this.ParsecFail() = run parsec "555-555-Narwhal"

File: Benchmarks/FailureComparison.cs
using System;
using System.Text.Patterns;
using System.Text.RegularExpressions;
using BenchmarkDotNet.Attributes;

namespace Benchmarks {
	[ClrJob, CoreJob, CoreRtJob]
	[MemoryDiagnoser]
	public class FailureComparison {

		readonly static Pattern pattern = Pattern.Number * 3 & '-' & Pattern.Number * 3 & '-' & Pattern.Number * 4;

		[Benchmark]
		public Result PatternFail() => pattern.Consume("555-555-Narwhal");

		readonly static Regex regex = new Regex(@"\d{3}-\d{3}-\d{4}", RegexOptions.IgnoreCase);

		[Benchmark]

		public Match RegexFail() => regex.Match("555-555-Narwhal");
	}
}
